src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+            See diagram 1 in the paper for more info on the default strategy
+        output_attentions (:obj:`bool`, `optional`, defaults to :obj:`None`):
+            If set to ``True``, the attentions tensors of all attention layers are returned. See ``attentions`` under returned tensors for more detail.
+"""
+
+class ProphetNetPreTrainedModel(PreTrainedModel):
+    config_class = ProphetNetConfig
+    base_model_prefix = "model"
+
+    def _init_weights(self, module):
+        std = self.config.init_std
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+
+class LearnedPositionalEmbedding(nn.Embedding):
+    """
+    This module learns positional embeddings up to a fixed maximum size.
+    Padding ids are ignored by either offsetting based on padding_idx
+    or by setting padding_idx to None and ensuring that the appropriate
+    position ids are passed to the forward function.
+    """
+
+    def __init__(
+            self,
+            num_embeddings: int,
+            embedding_dim: int,
+            padding_idx: int,
+    ):
+        super().__init__(num_embeddings, embedding_dim, padding_idx)
+        self.onnx_trace = False
+
+    def forward(self, input, use_cache=False, positions=None):
+        """Input is expected to be of size [bsz x seqlen]."""
+        assert (
+            (positions is None) or (self.padding_idx is None)
+        ), "If positions is pre-computed then padding_idx should not be set."
+
+        if positions is None:
+            if use_cache:
+                # positions is the same for every token when decoding a single step
+                # Without the int() cast, it doesn't work in some cases when exporting to ONNX
+                positions = input.data.new(1, 1).fill_(int(self.padding_idx + input.size(1)))
+            else:
+                mask = input.data.ne(self.padding_idx).int()
+                positions = ( torch.cumsum(mask, dim=1).type_as(mask) * mask ).long() + self.padding_idx
+            real_positions = positions
+        else:
+            real_positions = positions
+        return super().forward(positions), real_positions
+
+    def max_positions(self):
+        """Maximum number of supported positions."""
+        if self.padding_idx is not None:
+            return self.num_embeddings - self.padding_idx - 1
+        else:
+            return self.num_embeddings
+
+    def _forward(self, positions):
+        return super().forward(positions)
+
+def LayerNorm(normalized_shape, eps=1e-5, elementwise_affine=True, export=False):
```

Comment:
This is very different to our usual design...having upper case functions is very unexpected for a function (even though the function instantiates a class...). I think a better way here is to do the following at the very top of the code:

```
try:
    from apex.normalization import FusedLayerNorm
    ProphetNetLayerNorm  = FusedLayerNorm
except ImportError:
    ProphetNetLayerNorm = torch.nn.LayerNorm
```

then the layer norm can be instantiated normally with:

```ProphetNetLayerNorm(normalized_shape, eps, elementwise_affine)```

At the moment `export` is always set to `False` when calling `LayerNorm`  and the other two parameters are also not used...should we put `eps` and `elementwise_affine` in the config ? Do we need `export`? Or could we deleted it?

Reply:
Maybe import LayerNorm from another model is a good idea?

Reply:
like Bart, which does exactly this :)